cmd/kubeadm/app/cmd/upgrade/node.go

```diff
@@ -138,7 +142,12 @@ func newNodeData(cmd *cobra.Command, args []string, options *nodeOptions) (*node
 	// Fetches the cluster configuration
 	// NB in case of control-plane node, we are reading all the info for the node; in case of NOT control-plane node
 	//    (worker node), we are not reading local API address and the CRI socket from the node object
-	cfg, err := configutil.FetchInitConfigurationFromCluster(client, os.Stdout, "upgrade", !isControlPlaneNode)
```

Comment:
passing --config to upgrade open a can of worms.
ideally we want to fetch the one from the cluster and deprecate --config or rethink its scope.

Reply:
How is the ` upgrade node` command going to get the  appropriate `NodeRegistration.KubeletExtraArgs` when `configutil.FetchInitConfigurationFromCluster` only checks in kubeadm-config configmap? in most cases, this kubeadm-config configmap won't contain `NodeRegistration.KubeletExtraArgs`

Reply:
@neolit123 
I agree with @boluisa and actually myself filed https://cre.heptio.com/hc/requests/4666 ticket recently about the same issue. Our use case was as follows:
1. Essential PKS was renamed to Tanzu Kubernetes Grid
2. Our previous internal docker images repo had the name "essential-pks" in it, so we wanted to change to a new docker images repo that would now have "tanzu-kubernetes-grid" in the name.
3. We are not specifying NodeRegistration.KubeletExtraArgs in the config file, but are setting a new imageRepository in the config file and relying on the   https://github.com/kubernetes/kubernetes/pull/70603 change to automatically set the "--pod-infra-container-image" argument in /var/lib/kubelet/kubeadm-flags.env.
4. While imageRepository is in fact present in the `kubeadm-config` configmap, kubeadm upgrade commands are still currently not updating the /var/lib/kubelet/kubeadm-flags.env based on the new value of imageRepository.

The workaround that we implemented for now is to run `kubeadm init phase kubelet-start` prior to running the `kubeadm upgrade` command, but personally I believe it would have been super nice if the `kubeam upgrade` commands always automatically updated  /var/lib/kubelet/kubeadm-flags.env as this change is actually trying to do.

Reply:
yeah, we've seen other requests about that too.
--config on upgrade however is something that we want to move away from in the long term.
i has hidden risks in terms of cluster re-conf.

this is the issue in the k/kubeadm repository for the "reconf" feature that we want to add eventually:
https://github.com/kubernetes/kubeadm/issues/970

until then phases or manual edits are the recommended workarounds.

Reply:
I know it's a small detail about the can of worms, but regarding the node name one thing we are doing (at our SUSE tooling) is reading the `/etc/machine-id` locally. By listing the nodes, you can find the one with the `.nodeInfo.machineID` and see which one matches, so then you just use `.metadata.name` so it's not possible to break that bit.

That being said, I wholeheartedly agree that the reconf story should be something of its own, and the upgrade story locked down to avoid misconfigurations and upgrade failures. If you are reconfiguring the cluster you are aware of possible breakages, but upgrading the cluster should be a safe operation where we should try to protect the user from avoidable mistakes.